### ISSUE TITLE
can you share links to learn about bigartm and how it makes topic modeling

### ISSUE THREAD

**Issue by Sandy4321** [kellesett/probabilistic-topic-models-task#4]
can you share links to learn about bigartm and how it makes topic modeling please

**Comment by kellesett** [comment]
http://docs.bigartm.org/en/stable/index.html - this is only available source about bigartm application
If you are interested in some theory, I recommend [this course](http://bit.ly/2EGWcjA) (you will have to handle russian language)

**Comment by kellesett** [comment]
Also, as I know, you can start with PLSA and LDA methods, which are more common. BigArtm is quite complicated and provide much more features)

**Comment by kellesett** [comment]
Can you share why this method get your attention?

**Comment by Sandy4321** [comment]
I want to learn topic modeling

**Comment by Sandy4321** [comment]
http://www.machinelearning.ru/wiki/index.php?title=%D0%92%D0%B5%D1%80%D0%BE%D1%8F%D1%82%D0%BD%D0%BE%D1%81%D1%82%D0%BD%D1%8B%D0%B5_%D1%82%D0%B5%D0%BC%D0%B0%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B5_%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D0%B8_(%D0%BA%D1%83%D1%80%D1%81_%D0%BB%D0%B5%D0%BA%D1%86%D0%B8%D0%B9,_%D0%9A.%D0%92.%D0%92%D0%BE%D1%80%D0%BE%D0%BD%D1%86%D0%BE%D0%B2)

looks great but too long. I try to learn main ideas and how to use it

**Comment by Sandy4321** [comment]
i created cool material to explain main equation behind topic modeling
https://chatgpt.com/share/679152a7-4654-8000-bbe4-a5c5c09bd7b6

**Comment by Sandy4321** [comment]
i tried to understand 

![Image](https://github.com/user-attachments/assets/52ea44f9-afc7-42fe-b468-4bacbe0a8f37)

**Comment by Sandy4321** [comment]
https://www.coursera.org/learn/unsupervised-learning

**Comment by Sandy4321** [comment]
https://chatgpt.com/share/679152a7-4654-8000-bbe4-a5c5c09bd7b6

**Comment by Sandy4321** [comment]
https://chatgpt.com/share/679152a7-4654-8000-bbe4-a5c5c09bd7b6

**Comment by Sandy4321** [comment]
import random
import numpy as np
import math

##############################################################################
# 1) Synthetic Text Generation
##############################################################################

def generate_synthetic_text_data(
    D=10,           # Number of documents
    topics_count=5, # Total number of topics
    words_per_topic=8, 
    doc_min_topics=2,
    doc_max_topics=4,
    doc_min_length=20,
    doc_max_length=40,
    seed=42
):
    """
    Simulate a corpus of text documents using 'topics_count' synthetic topics.
    
    1) Create a list of 'topics_count' topics, each with 'words_per_topic' distinct words.
    2) For each of D documents:
       - Randomly choose T topics (T in [doc_min_topics, doc_max_topics]).
       - Sample total [doc_min_length .. doc_max_length] words from those T topics.
       - Shuffle them to mimic a real text order and join into a string.
    
    Returns:
      doc_texts: List of document strings
      topics:    List of lists, each sub-list is the word set for that topic
    """
    random.seed(seed)
    
    # Create synthetic topics
    topics = []
    for t in range(topics_count):
        topic_words = [f"topic{t}_word{w}" for w in range(words_per_topic)]
        topics.append(topic_words)
    
    doc_texts = []
    
    for d in range(D):
        # Pick how many topics for this doc
        T = random.randint(doc_min_topics, doc_max_topics)
        chosen_topics = random.sample(range(topics_count), T)
        
        # Decide how many total words in this doc
        doc_length = random.randint(doc_min_length, doc_max_length)
        
        # We'll roughly split doc_length among the chosen topics
        words_so_far = []
        for t_idx in chosen_topics:
            n_t = doc_length // T  # naive division
            for _ in range(n_t):
                chosen_word = random.choice(topics[t_idx])
                words_so_far.append(chosen_word)
        
        # If not perfectly divisible, fill leftover
        while len(words_so_far) < doc_length:
            leftover_topic = random.choice(chosen_topics)
            words_so_far.append(random.choice(topics[leftover_topic]))
        
        # Shuffle words for a more "natural" ordering
        random.shuffle(words_so_far)
        
        # Join into a single string
        doc_text = " ".join(words_so_far)
        doc_texts.append(doc_text)
    
    return doc_texts, topics

def tokenize_documents(doc_texts):
    """
    Very simple tokenization: split on whitespace.
    Returns a list of lists of tokens.
    """
    doc_tokens = []
    for text in doc_texts:
        tokens = text.split()
        doc_tokens.append(tokens)
    return doc_tokens

def build_doc_term_matrix(doc_tokens):
    """
    Build a document-term matrix from tokenized documents.
    Returns:
      doc_term_matrix (D x V) and vocabulary (list of unique words).
    """
    # Gather all words
    all_words = set()
    for tokens in doc_tokens:
        all_words.update(tokens)
    
    vocabulary = sorted(all_words)
    word_index = {w: i for i, w in enumerate(vocabulary)}
    
    D = len(doc_tokens)
    V = len(vocabulary)
    doc_term_matrix = np.zeros((D, V), dtype=int)
    
    for d, tokens in enumerate(doc_tokens):
        for w in tokens:
            w_idx = word_index[w]
            doc_term_matrix[d, w_idx] += 1
    
    return doc_term_matrix, vocabulary

##############################################################################
# 2) Topic Model: Perplexity & Regularized Online EM
##############################################################################

def compute_perplexity(doc_term_matrix, Theta, Phi):
    """
    Compute perplexity for a doc-term matrix, doc-topic (Theta), topic-word (Phi).
    """
    D, V = doc_term_matrix.shape
    total_words = doc_term_matrix.sum()
    log_likelihood = 0.0
    
    for d in range(D):
        for w in range(V):
            count = doc_term_matrix[d, w]
            if count > 0:
                p_w_d = np.dot(Theta[d], Phi[:, w])  # sum_k Theta[d,k]*Phi[k,w]
                p_w_d = max(p_w_d, 1e-12)
                log_likelihood += count * math.log(p_w_d)
    
    perp = math.exp(-log_likelihood / total_words)
    return perp

def apply_decorrelation_penalty(Phi, alpha=0.01):
    """
    Gradient-like step to reduce dot-product overlap among topics.
    Penalty: sum_{k<k'} (dot product of Phi[k], Phi[k']), i.e. sum_w Phi[k,w]*Phi[k',w].
    
    Update:
      Phi[k,w] -= alpha * (sum_{k' != k} Phi[k',w])
    Then clip negatives and re-normalize each topic row.
    """
    K, V = Phi.shape
    grad_penalty = np.zeros_like(Phi)
    
    # For each k, sum up all other topics' probabilities for each w
    for k in range(K):
        sum_others = np.sum(Phi[[i for i in range(K) if i != k]], axis=0)
        grad_penalty[k] = sum_others
    
    # Subtract penalty
    Phi_new = Phi - alpha * grad_penalty
    
    # Clip negatives
    Phi_new[Phi_new < 0] = 0.0
    
    # Re-normalize each row
    row_sums = Phi_new.sum(axis=1, keepdims=True)
    row_sums[row_sums < 1e-12] = 1e-12
    Phi_new /= row_sums
    
    return Phi_new

def regularized_online_em_topic_modeling(
    doc_term_matrix,
    K,
    num_epochs=10,
    batch_size=10,
    learning_rate=0.7,
    decorrel_strength=0.01
):
    """
    Online EM for a mixture-of-unigrams model, with a decorrelation penalty 
    that reduces overlap among topic-word rows (Phi).
    
    doc_term_matrix: (D, V)
    K: number of topics
    num_epochs: number of passes over the data
    batch_size: mini-batch size
    learning_rate: base step size (decays ~ 1/sqrt(epoch))
    decorrel_strength: alpha step for the dot-product penalty
    
    Returns:
      Phi:    (K, V) final topic-word matrix
      Thetas: (D, K) doc-topic distribution for each doc
    """
    D, V = doc_term_matrix.shape
    
    # 1) Initialize global Phi randomly
    Phi = np.random.rand(K, V)
    Phi /= Phi.sum(axis=1, keepdims=True)
    
    # 2) We'll maintain doc-topic distributions
    Thetas = np.zeros((D, K))
    
    doc_indices = np.arange(D)
    num_batches = int(np.ceil(D / batch_size))
    
    for epoch in range(num_epochs):
        # Shuffle documents each epoch
        np.random.shuffle(doc_indices)
        
        # Decaying learning rate
        rho = learning_rate / np.sqrt(epoch + 1)
        
        for b in range(num_batches):
            start_idx = b * batch_size
            end_idx = min((b+1)*batch_size, D)
            batch_docs = doc_indices[start_idx:end_idx]
            current_size = len(batch_docs)
            
            # Local E-step:
            gamma = np.zeros((current_size, K))
            n_hat = np.zeros((K, V))
            
            for i, d in enumerate(batch_docs):
                # Approx doc-topic distribution:
                # gamma[d,k] ~ prod_w (Phi[k,w])^N[d,w]
                log_vals = np.zeros(K)
                for k_ in range(K):
                    val = 0.0
                    for w in range(V):
                        cnt = doc_term_matrix[d, w]
                        if cnt > 0:
                            pkw = max(Phi[k_, w], 1e-12)
                            val += cnt * math.log(pkw)
                    log_vals[k_] = val
                # exponentiate and normalize
                exp_vals = np.exp(log_vals - log_vals.max())
                gamma_sum = exp_vals.sum()
                if gamma_sum < 1e-12:
                    gamma_sum = 1e-12
                gamma[i] = exp_vals / gamma_sum
                
                # Accumulate n_hat
                for k_ in range(K):
                    n_hat[k_] += gamma[i, k_] * doc_term_matrix[d]
            
            # Store doc-topic in Thetas
            for i, d in enumerate(batch_docs):
                Thetas[d] = gamma[i]
            
            # Global M-step: update Phi
            scaling = float(D) / current_size
            n_hat_scaled = n_hat * scaling
            
            new_Phi = np.zeros_like(Phi)
            for k_ in range(K):
                row_sum = n_hat_scaled[k_].sum()
                row_sum = max(row_sum, 1e-12)
                new_Phi[k_] = n_hat_scaled[k_] / row_sum
            
            # Combine old Phi and new_Phi with step size rho
            Phi = (1 - rho)*Phi + rho*new_Phi
            
            # Apply decorrelation penalty
            alpha = decorrel_strength * rho
            Phi = apply_decorrelation_penalty(Phi, alpha=alpha)
        
        # Check perplexity each epoch
        perp = compute_perplexity(doc_term_matrix, Thetas, Phi)
        print(f"Epoch {epoch+1}/{num_epochs}, Perplexity = {perp:.2f}")
    
    return Phi, Thetas

##############################################################################
# 3) Putting It All Together
##############################################################################

def main():
    # 1) Generate synthetic text documents with 5 topics
    #    Each document uses 2-4 of these topics, with length 20-40 words
    D = 20
    topics_count = 5
    words_per_topic = 8
    doc_texts, synthetic_topics = generate_synthetic_text_data(
        D=D,
        topics_count=topics_count,
        words_per_topic=words_per_topic,
        doc_min_topics=2,
        doc_max_topics=4,
        doc_min_length=20,
        doc_max_length=40,
        seed=0
    )
    
    # Print example synthetic topics
    print("\n=== Synthetic Topics (Distinct Word Groups) ===")
    for t_idx, tw in enumerate(synthetic_topics):
        print(f"Topic {t_idx}:", tw)
    
    # Print example documents
    print("\n=== Example Documents (Simulated Text) ===")
    for i, doc in enumerate(doc_texts[:3]):
        print(f"Doc {i}:\n  {doc}\n")
    
    # 2) Tokenize
    doc_tokens = tokenize_documents(doc_texts)
    
    # 3) Build doc-term matrix
    doc_term_matrix, vocabulary = build_doc_term_matrix(doc_tokens)
    print("Vocabulary size:", len(vocabulary))
    print("Doc-term matrix shape:", doc_term_matrix.shape)
    
    # 4) Run Regularized Online EM
    K = 5             # We'll learn 5 topics
    num_epochs = 10
    batch_size = 5
    learning_rate = 0.7
    decorrel_strength = 0.1  # how strongly we push topics apart
    
    print("\n=== Training Topic Model (Online EM + Decorrel) ===")
    Phi_est, Theta_est = regularized_online_em_topic_modeling(
        doc_term_matrix,
        K=K,
        num_epochs=num_epochs,
        batch_size=batch_size,
        learning_rate=learning_rate,
        decorrel_strength=decorrel_strength
    )
    
    # 5) Final Perplexity
    final_perp = compute_perplexity(doc_term_matrix, Theta_est, Phi_est)
    print(f"\nFinal Perplexity: {final_perp:.2f}\n")
    
    # 6) Inspect top words in each learned topic
    print("=== Learned Topics (Top Words) ===")
    for k_ in range(K):
        row = Phi_est[k_]
        top_words_idx = np.argsort(-row)[:5]
        top_words = [(vocabulary[i], row[i]) for i in top_words_idx]
        print(f"Topic {k_}:", top_words)

if __name__ == "__main__":
    main()

**Comment by kellesett** [comment]
I don’t think that’s a good idea to insert a lot of text in this issue but if you have some specific questions - you’re welcome. Yes, course on probabilistic topic modeling is quite big, but it contains a lot of theory, optimization details and research results - you can retrieve only what you need
And you also can try to find some materials about LDA and PLSA

**Comment by Sandy4321** [comment]
if you have some specific questions - you’re welcome

so how to install bigartm on windows
can you share changed code what works on windows ?

**Comment by kellesett** [comment]
Have you tried to follow installation instructions from official site with docs?

**Comment by Sandy4321** [comment]
not works for windows
